### PR TITLE
Permissions refactor

### PR DESCRIPTION
**Pull request by afinch7**
Refactored permissions to be assignable on a per-isolate basis, and added a fix for #1858 to `op_fetch_module_meta_data`.

### CODE REVIEW

File: src/workers.rs [hunk 1]
```diff
@@ -54,6 +58,7 @@ impl Worker {
 pub fn spawn(
```
Comment (ry) [inline src/workers.rs]: Please add some triple slash docs here.

What does it mean when permissions is None? How is that different than having the Permission struct present but with all the fields set to false?

File: src/isolate.rs [hunk 2]
```diff
@@ -195,6 +169,7 @@ impl Isolate {
     snapshot: libdeno::deno_buf,
     state: Arc<IsolateState>,
     dispatch: Dispatch,
+    permissions: Option<DenoPermissions>,
```
Comment (ry) [inline src/isolate.rs]: Please add some triple slash docs here.

What does it mean when permissions is None? How is that different than having the Permission struct present but with all the fields set to false?

File: src/isolate.rs [hunk 3]
```diff
@@ -208,6 +183,8 @@ impl Isolate {
     let libdeno_isolate = unsafe { libdeno::deno_new(config) };
     // This channel handles sending async messages back to the runtime.
     let (tx, rx) = mpsc::channel::<(usize, Buf)>();
+    let isolate_perms =
+      Arc::new(permissions.unwrap_or(DenoPermissions::new(&state.flags)));
```
Comment (ry) [inline src/isolate.rs]: I think this is strange that the permissions come from state.flags or from this extra argument. I'd rather explicitly have the caller create the permission struct and always pass it, than this implicit logic.

**Comment by afinch7** [comment]
I made the requested changes by making permissions a required parameter, and added 3 different tests for `op_fetch_module_meta_data`.

File: src/ops.rs [hunk 4]
```diff
@@ -365,11 +374,28 @@ fn op_fetch_module_meta_data(
   let specifier = inner.specifier().unwrap();
   let referrer = inner.referrer().unwrap();
 
-  assert_eq!(state.dir.root.join("gen"), state.dir.gen, "Sanity check");
+  if !isolate.permissions.allow_read.load(Ordering::SeqCst) {
```
Comment (bartlomieju) [inline src/ops.rs]: If I understand correctly we only want to give "compiler" isolate access to this op. Some comment would be welcomed here, as all other ops prompt for missing permissions.

Comment (bartlomieju) [reply src/ops.rs]: Yeah I get it, it's just kind of "special" case - all other ops prompt for permission. I just wanted one-line comment on this special case :)

**Pull request merged**